### PR TITLE
SP800-108 CTR HMAC

### LINKED ISSUE #65577: API proposal for SP800-108-CTR-HMAC cryptographic algorithm
## Motivation

A **key derivation function (KDF)** is a building block operation of any cryptosystem, alongside cipher algorithms and MAC algorithms. Within Microsoft, many of our partners use SP800-108-CTR-HMAC, and it is commonly recommended by the Crypto Board. (See [NIST SP 800-108](https://csrc.nist.gov/publications/detail/sp/800-108/final), §5.1.) ASP.NET's DataProtection API [also uses this KDF](https://docs.microsoft.com/aspnet/core/security/data-protection/implementation/subkeyderivation#additional-authenticated-data-and-subkey-derivation) under the covers.

.NET does not currently provide a managed wrapper around the SP800-108-CTR-HMAC algorithm, forcing all consumers to implement it themselves from spec. Representatives from the Crypto Board have asked us to create a supported .NET implementation that can be used by our mutual customers. Like the work we did to add _System.IO.Hashing_ (see https://github.com/dotnet/runtime/pull/53623), we should ensure we produce a netstandard2.0-consumable package for customers who target .NET Framework.

## API Proposal

```cs
namespace System.Security.Cryptography;

public sealed class SP800108HmacCounterKdf : IDisposable
{
    //
    // CREATE A REUSABLE INSTANCE WRAPPED AROUND A FIXED KDK
    //
    public SP800108HmacCounterKdf(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm);
    public SP800108HmacCounterKdf(byte[] key, HashAlgorithmName hashAlgorithm);

    //
    // REUSABLE INSTANCE METHODS
    //
    public byte[] DeriveKey(byte[] label, byte[] context, int derivedKeyLengthInBytes);
    public byte[] DeriveKey(ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, int derivedKeyLengthInBytes);
    public byte[] DeriveKey(string label, string context, int derivedKeyLengthInBytes);
    public byte[] DeriveKey(ReadOnlySpan<char> label, ReadOnlySpan<char> context, int derivedKeyLengthInBytes);
    public void DeriveKey(ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, Span<byte> destination);
    public void DeriveKey(ReadOnlySpan<char> label, ReadOnlySpan<char> context, Span<byte> destination);

    //
    // STATIC ONE-SHOTS
    //
    public static byte[] DeriveBytes(byte[] key, HashAlgorithmName hashAlgorithm, byte[] label, byte[] context, int derivedKeyLengthInBytes);
    public static byte[] DeriveBytes(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, int derivedKeyLengthInBytes);
    public static byte[] DeriveBytes(byte[] key, HashAlgorithmName hashAlgorithm, string label, string context, int derivedKeyLengthInBytes);
    public static byte[] DeriveBytes(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm, ReadOnlySpan<char> label, ReadOnlySpan<char> context, int derivedKeyLengthInBytes);
    public static void DeriveBytes(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, Span<byte> destination);
    public static void DeriveBytes(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm, ReadOnlySpan<char> label, ReadOnlySpan<char> context, Span<byte> destination);

    public void Dispose();
}
```

## Terminology

* __key derivation key (KDK)__: the master key from which subkeys (derived keys) will be generated; taken as the _key_ parameter in these APIs.
* __pseudo-random function (PRF)__: the HMAC algorithm which is used to generate the subkey; taken as the _hashAlgorithm_ parameter in these APIs.
* __label__ and __context__: a descriptive tuple which changes how the subkey is generated and allows generation of multiple unique subkeys from a single KDK. (ASP.NET's DataProtection APIs combine these into a single concept called a [_purpose string_](https://docs.microsoft.com/aspnet/core/security/data-protection/consumer-apis/purpose-strings).)

## Usage notes

> __Security note:__ Callers __must__ ensure the KDK input to these APIs consists of strong entropy. If the input key contains low entropy (e.g., a user-specified password), a key-stretching KDF like PBKDF2 (via `RFC2898DeriveBytes`) must be used instead. This is the caller's responsibility; our APIs will not fail if an empty or low entropy KDK is provided.

Per spec, the _hashAlgorithm_ parameter __must__ be one of: `SHA1`, `SHA256`, `SHA384`, or `SHA512`. (We wrap it within an HMAC construction.) Any other algorithm selection will result in `ArgumentException`.

## Discussion

Two patterns of APIs are provided. The easiest pattern is to specify the desired length of the subkey, and the system will allocate a new array and return it to the caller. If the caller wishes to be in full control over memory allocation, a "fill the entire output buffer" overload is available. For each of these three patterns, instance methods and static one-shots are available. If the `Secret<T>` type is introduced in the future (see https://github.com/dotnet/designs/pull/147), we should introduce overloads which take the kdk as `Secret<byte>` and which return the derived subkey as a new `Secret<byte>`, which follows ASP.NET's existing usage.

Since the _label_ and _context_ parameters could be binary or text, we should for convenience allow the caller to pass in char spans / strings. These overloads will convert the input to UTF-8 and call the binary overloads. If the input contains malformed UTF-16 text, `ArgumentException` is thrown. Callers should take care not to allow embedded nulls within the _label_ parameter. However, our APIs will not check for embedded nulls and will not fail should they be encountered.

The KDK (_key_) parameter is expected to be a cryptographic key with strong entropy. This is raw binary data, not textual data, so there is no need for UTF-16 overloads for the _key_ parameter.

Unlike `RFC2898DeriveBytes`, the instance methods proposed here are stateless. Constructing an instance and calling `DeriveBytes` over and over will - assuming constant _label_ and _context_ inputs and subkey lengths - produce the same output byte-for-byte. If the caller wishes to generate multiple subkeys from a single KDK, they should: (a) generate a single long subkey and slice pieces off as needed; or (b) vary the _label_ or _context_ parameters and run the operation again.

The type is sealed to match the designs of the recent cryptography primitives `AesGcm` and `ChaCha20Poly1305`. The SP800-108-CTR-HMAC algorithm is itself an opaque primitive and no extensibility points are considered per spec. Once the KDK and PRF are captured by the constructor, there is no public API to retrieve them.

## Thread safety

Instances of this type shall be thread-safe for multiple callers. Instance methods shall not mutate internal state or persist state between calls. This is because applications might retain a single KDK for multiple operations, and they'll need to generate subkeys on-demand from arbitrary threads. (ASP.NET relies on this behavior.)

## Packaging

Like _System.IO.Hashing_, we have considerable interest in this API from consumers on downlevel runtimes. We should create a new netstandard2.0-compatible package _System.Security.Cryptography.SP800108_ and expose the APIs from there. If needed, this package can have a different implementation on .NET 6+ or .NET 7, including any appropriate type-forwarding.

## Alternative proposals

The names _SP800108DeriveBytes_ and _SP800108CtrHmac_ were proposed, with the former mirroring the _RFC2898DeriveBytes_ naming. However, the _\*Hmac_ suffix was deemed to be misleading, since this algorithm isn't itself an HMAC. And leaving the "counter / HMAC" description out entirely isn't descriptive enough of what the API does.

This type would be one of the few thread-safe types in the _System.Security.Cryptography_ namespace. If this asymmetry is not desired, we could consider making the type `ICloneable`, which would allow spawning a new instance with its own dedicated lifetime, and with the captured KDK and PRF duplicated. We could potentially get away with saying that callers who need thread safety should use the static one-shots. However, that would impact the performance of ASP.NET, which has its own implementation of this type which makes thread-safety guarantees.

[The existing type `HKDF`](https://apisof.net/catalog/0431a92b-ff1a-f4e8-c1fd-6d58a0d259e0) takes its _salt_ and _info_ parameters at the end of the APIs because they're considered optional. I opt not to do that here, placing the _label_ and _context_ parameters upfront as required inputs (which can be empty). This is because SP800-108 very strongly recommends callers pass meaningful data for these parameters (see §§7.5 - 7.6), which allows a single KDK to generate an arbitrary number of subkeys. If the application instead has a fixed number of uses, they can use a static one-shot and pass empty arrays / strings / spans for these APIs, slicing subkeys as needed.

### PR DESCRIPTION
**Pull request by vcsjones**
This implements SP800-108 CTR HMAC as a package. In .NET 8, this is an inbox component. For down-level platforms, `Microsoft.Bcl.Cryptography` was introduced as a package for .NET Standard and .NET Framework.

Some implementation notes.

1. This at runtime decides if the platform is Windows 8+. If it is, you get the CNG implementation. If it isn't, you get the Managed implementation.
2. Both the CNG and Managed implementation support NetCoreApp and .NET Standard. This results in some backward desirable behavior. The .NET Standard implementation prefers working with `byte[]` for inputs because of `IncrementalHash`. .NET Core prefers working with spans. Because of this, we avoid the `byte[]` APIs deferring to `{ReadOnly}Span<byte>` APIs. Otherwise, we will go from `byte[]` -> span -> `byte[]`. This is true not just for public APIs that accept input, but also for places we need to do conversions, like from `string` to `byte[]`. In .NET Standard, we always rent from the pool and avoid doing work on the stack, whereas .NET Core can use the stack.
3. This class is **thread safe** for all implementations. This includes instance methods. Instances methods should either a) product the correct result or b) throw `ObjectDisposedException`. We get this for free with safe handles and the CNG implementation. For the managed implementation, the key is reference counted. When the reference count reaches zero, the key is zeroed and attempts to acquire the key afterward throw an `ObjectDisposedException`.
4. The test data draws from a few different places (explained in the tests) such as RFCs and other implementations like SymCrypt. While NIST has CAVP test vectors, those tests use 800-108 CTR HMAC in a mode that our public API does not support. All of the test vectors assume you have total control over the input into the PRF, whereas our public API forces the input to be `label || 0x00 || context`. So the NIST test vectors are unusable for tests.

Closes #65577

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks.

We really need a new package/assembly for this one type?  We can't, for example, just squint and say it's related to hashing so it can go in System.IO.Hashing?

**Comment by vcsjones** [comment]
> We really need a new package/assembly for this one type?

There is internal (from Microsoft) interest in this capability for .NET Framework, and the API proposal suggested this as a package name.

> We can't, for example, just squint and say it's related to hashing so it can go in System.IO.Hashing?

My 2c: that's... quite a bit of a squint. That package is specifically documented as "non cryptographic hashing" in its README. This API is 1. cryptographic and 2. doesn't do hashing, strictly speaking (though the PRF is HMAC in this one case).

Thinking about it more, I don't really see a good "home" for this type in an existing package, so a new package makes sense to me.

Though I will defer to y'all.

**Comment by stephentoub** [comment]
If we really lack a good existing package/assembly for it, then we should think hard about what to name this one so we don't need to create yet another the next time we need to ship another cryptographic algorithm downlevel.

**Comment by vcsjones** [comment]
>  we should think hard about what to name this one so we don't 

@terrajobst @bartonjs @GrabYourPitchforks do y'all have any thoughts here?

**Comment by vcsjones** [comment]
> System.Security.Cryptography.NETStandard

I would probably avoid calling anything `NETStandard`, but I don't have a big preference for what the generic name is.

I do have a bit of a preference for using a generic name at all. One thing I don't really want is a big invitation / debate on what should be in `NETStandard` or whatever going forward. Let's say, of the sake of argument, HKDF doesn't in .NET today. Would we put it in this `Extensions` or `NETStandard` package? What about `CertificateRevocationListBuilder`?

I think there is clear evidence that SP800108 needs to be made available in downlevel targets. What's not clear to me is why everything else couldn't argue that criteria.

> I don't remember if a super-generic bucket like that was proposed and shot down, or if I'm striking gold here.

It was discussed in the API review here. https://youtu.be/bDTLx5SXusY?t=5714 (1:35:20 in).

My take away from that discussion a slight preference to avoid a "dumping ground" package and "fine with another one off package".

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs:90]: Consider pulling this out as a const so that it can be updated everywhere by changing a single line. Also consider what a common label \& context length might be. A scratch buffer of 128 bytes means that any source span \>= 43 chars will hit the crypto pool.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs:11]: I wonder if this should be moved to a common helper file and shared across projects. System.Text.Json has similar code, as shown below. (And both this file and the JSON copy have subtle bugs when handling zero-length source \& destination, which doesn't matter _here_ but we should probably address nonetheless if we're going to get in the habit of copying this code.)

https://github.com/dotnet/runtime/blob/c6d6b19da5427ae78eab241d2af596a3320a5e5b/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L247-L258

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs:73]: The only difference between the `byte[]` and `ReadOnlySpan<byte>` versions is the HashOneShot method call.  Can we collapse these into one big body via something like

```
_keyHandle = OpenKeyHandle(key, hashAlgorithm);

...

private static SafeSomethingHandle OpenKeyHandle(byte[] key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key.AsSpan());
}

private static SafeSomethingHandle OpenKeyHandle(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key);
}
```

Then the Core just gets to funnel down the "don't pass null", and is common for both paths.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs:11]: The subtle bug being... what? When a BOM needs to get written?

Comment (GrabYourPitchforks) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs:11]: > The subtle bug being... what? When a BOM needs to get written?

The `GetChars` method can't handle null inputs or outputs, and _in the general case_ there's no guarantee that a zero-length input produces a zero-length output. Doesn't matter for UTF-8, but I'm seeing this code duplicated in a few places and we may as well just write it once and get it correct from the beginning.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:45]: Nit: For internal use, prefer abstract base classes instead of interfaces where possible. Just makes the codegen a tiny bit cleaner.

(If you have other reasons for preferring an interface, don't fret too much on this.)

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:215]: Nit - grammar: _contain_ -\> _contains_. (Same comment elsewhere.)

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:219]: Nit - grammar: _accept_ -\> _accepts_

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:92]: Technically this is an observable change: calling `Dispose` multiple times could cause the key to be zeroed while an operation is taking place, which would cause the `DeriveBytes` operation to return an _incorrect_ value (as opposed to throwing an exception). A standard `SafeHandle` implementation works around this by special-casing `Dispose` to only decrement the ref count once.

You could work around it with something like this:

```cs
int _disposed = 0;
public void Dispose()
{
    if (Interlocked.Exchange(ref _disposed, 1) == 0) { ReleaseKey(); }
}
```

It also means that you don't need a loop within `ReleaseKey`. The logic becomes a whole lot easier.

```cs
private void ReleaseKey()
{
    int newCount = Interlocked.Decrement(ref _keyReferenceCount);
    Debug.Assert(newCount >= 0, "How did this go negative?");
    if (newCount == 0) { ZeroKey(); }
}
```

`SafeHandle` has more complexity to keep the "was disposed?" bit _within_ the ref count integer, but that seems like overkill here.

**Comment by GrabYourPitchforks** [comment]
I've finished review of the prod logic - haven't gotten to tests yet. Some nits, and a race condition in `ReleaseKey` that we should address, but everything LGTM. Will try to get to the tests in a few days.

Thanks Kevin for getting to this! It's awesome to see it come together. :)

> I think there is clear evidence that SP800108 needs to be made available in downlevel targets. What's not clear to me is why everything else couldn't argue that criteria.

Your line of reasoning is valid. I'm seeing some internal requests for AES-GCM (and maybe AES-CCM?) to be made available downlevel as well since folks are clamoring for them on net4x.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:92]: Ohhhhhhhhhhhh good catch. This uncovered a few problems with the reference counting. This along with some asserts makes me feel better about the whole thing.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:45]: > If you have other reasons for preferring an interface

Originally I did some cute things with static abstracts on the interface, and later remembered that did not work on downlevel platforms, so no good reason.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs:11]: That example is not exactly the same. That one is doing `GetChars`. We are doing `GetBytes`. So we can't really consolidate these two specific examples. I can extract out helpers if you want, but in this particular case it won't result in de-duping anything.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs:73]: I tried this originally and it doesn't really improve much. In addition to the hashing, the clearing of the hashed key makes it difficult to mix and match `byte[]`, `Span<byte>` and `ReadOnlySpan<byte>`. As I worked on refactoring and fixing it, I pretty much just ended up right where I was again.

**Comment by vcsjones** [comment]
@GrabYourPitchforks I believe I have addressed all of your feedback thus far, with most of it actioned, some of it "maybe not". I appreciate your time thus far!

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs]: Labels are mostly driven by RFCs or are usually on the small side. Contexts... can really go anywhere in size. I ended up bumping it to 256 because we should have the stack space for it.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs]: > but should it be private?

Yes it should have.

**Comment by GrabYourPitchforks** [comment]
> That would probably have to be Windows-only or limited to .NETFramework. Supporting a true netstandard2.0 for GCM / CCM would be a huge multiplier in complexity.

Oh, it would definitely add complexity. But just yesterday I was in an email thread (which I think you were on as well!) where the writing on the wall looms ever larger.

Fortunately that's a problem for future us and not for this PR! 🫠

**Review by jeffhandley — DISMISSED** [review]
I know we've still been waiting for a re-review from @GrabYourPitchforks, but from what I can tell all previous comments have been addressed. @bartonjs, if this looks good to you, I propose we go ahead and merge this ahead of Preview 1. We can revisit the packaging topic if another idea emerges there, and if there's post-merge feedback from @GrabYourPitchforks, we can address it in a separate PR.

**Comment by stephentoub** [comment]
>  We can revisit the packaging topic if another idea emerges there

If we rename it, won't that mean we'll have another dead-end package on nuget forevermore?  It's not the end of the world, just, not ideal.

**Comment by jeffhandley** [comment]
> > We can revisit the packaging topic if another idea emerges there
> 
> If we rename it, won't that mean we'll have another dead-end package on nuget forevermore? It's not the end of the world, just, not ideal.

True. I rationalized it as: It'd only exist with `-preview` version(s) and we could unlist it while keeping it available for package restore).

Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?

**Comment by vcsjones** [comment]
> Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?

My 2 cents, it's my preferred choice.

**Review by jeffhandley — COMMENTED** [review]
Nit-picky: It looks like we always use lower-case `.netstandard.cs` and `.netcoreapp.cs` throughout the repo. Can you rename those files to lowercase the suffixes please?

**Comment by jeffhandley** [comment]
> > Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?
> 
> My 2 cents, it's my preferred choice.

@bartonjs is going to re-review, so we'll give him a chance to weigh in on that again. In anticipation of that being OK, I'm tempted to get a commit ready with that restructuring. Would you like to do that, @vcsjones; or would you like me to? (Either's good with me; I have time today).

**Comment by vcsjones** [comment]
> Would you like to do that, @vcsjones

I should be able to get that in over the next few hours.

**Comment by bartonjs** [comment]
If we want to propose a new package name, that needs to go back to API Review.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:48]: There's a bit of cave-man speech going on here.  I think the standard form I've seen is

```suggestion
        ///   Initializes a new instance of <see cref="SP800108HmacCounterKdf" /> using a specified key and HMAC algorithm.
```

(also applies to the next ctor)

**Review by bartonjs — DISMISSED** [review]

**Comment by bartonjs** [comment]
Since there's active discussion about renaming it, it should go back to API Review.  If this gets merged in the next couple of hours, it makes it into Preview 1... which we don't want to do if we're going to rename it literally tomorrow.  So, despite having signed off on the code, marked the PR as No Merge.

**Comment by vcsjones** [comment]
> it should go back to API Review

Okay. I'll address some nits but hold off on pushing the commit that does the rename until it comes back from API review.

File: src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:48]: SP800-108 good. Cave man speech bad!

**Comment by vcsjones** [comment]
@ViktorHofer added you to get input on the packaging for `Microsoft.Bcl.Cryptography`. The background is we added a new type called `SP800108HmacCounterKdf`. This is inbox in `System.Security.Cryptography`. For down level targets, that package was introduced to contain an implementation for .NET Framework and .NET Standard.

File: src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs
Comment (bartonjs) [inline src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs:172]: It probably only wants to be a `//` comment instead of xmldoc... but either way "ins" isn't English :smirk:

```suggestion
        /// Returns null if the platform is Windows 10+ and psuedo handles should be used.
```

File: src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (bartonjs) [inline src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:103]: Nit: Distracting trailing blank line

```suggestion
```

File: src/libraries/Microsoft.Bcl.Cryptography/ref/Microsoft.Bcl.Cryptography.csproj
Comment (bartonjs) [inline src/libraries/Microsoft.Bcl.Cryptography/ref/Microsoft.Bcl.Cryptography.csproj:9]: Alignment:

```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```

File: src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (bartonjs) [inline src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:29]: I was about to ask why keyBuffer is using gcnew here.  Then realized it's probably because IncrementalHash wants a precisely-sized array input.  Probably worth a comment.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdf.Windows.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdf.Windows.cs:13]: Adding `readonly` let's the Tier-N(>0) JIT treat it as a const.

```suggestion
        private static readonly bool s_isWindows8OrGreater = OperatingSystem.IsWindowsVersionAtLeast(6, 2);
```

**Review by bartonjs — APPROVED** [review]

File: src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs
Comment (vcsjones) [reply src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:30]: Yep. Added a comment.

**Comment by ericstj** [comment]
@vcsjones would this be interesting to add to https://github.com/dotnet/core/issues/8134?  If so please consider adding today before end of day.  Sorry for the late notice.

**Comment by vcsjones** [comment]
> would this be interesting to add to https://github.com/dotnet/core/issues/8134?

Eh, I think it's cool, but I don't think it has broad enough appeal, and demonstrating it is tricky.

**Pull request merged**
